### PR TITLE
Implement GetSymbolInfo for the identifier in a property pattern.

### LINKED ISSUE #9284: SymbolInfo for bad property in a property pattern should contain candidate symbols
The `SymbolInfo` returned when there is a bad property in a property pattern (e.g. because it is ambiguous, inaccessible, `static`, etc) should contain any properties that are (improperly) referenced.

See, for example, the test `AmbiguousNamedProperty` in `PatternMatchingTests`.

### LINKED ISSUE #26139: SemanticModel.GetSymbolInfo not working on property pattern names
This is a bug in the `features/recursive-patterns` branch.

/cc @CyrusNajmabadi

### LINKED ISSUE #26607: Semantic model not working for property sub pattern name.
```c#
class C
{
    void M2()
    {
        object o = null;
        _ = o switch { Blah { X: Y } => 0, _ => 0 }; // try GetSymbolInfo on either X or the entire subpattern element
    }

    class Blah
    {
        public int X;
    }
}
```

Expected: get back the field symbol for Blah.X.  
Action: get back no symbol.

Note that Y is not defined.  But that should not impact if we return X or not.

### PR DESCRIPTION
**Pull request by gafter**
Fixes #9284
Fixes #26607
Fixes #26139

@cston @agocke Please review this pattern-matching change to implement GetSymbolInfo for subpatterns.
/cc @dotnet/roslyn-compiler

### CODE REVIEW

File: src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests3.cs [hunk 1]
```diff
@@ -0,0 +1,369 @@
...
+using Microsoft.CodeAnalysis.Test.Utilities;
+using Roslyn.Test.Utilities;
+using Xunit;
+
+namespace Microsoft.CodeAnalysis.CSharp.UnitTests
+{
+    [CompilerTrait(CompilerFeature.Patterns)]
+    public class PatternMatchingTests3 : PatternMatchingTestBase
+    {
+        void AssertEmpty(SymbolInfo info)
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests3.cs]: private

File: src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests3.cs [hunk 2]
```diff
@@ -0,0 +1,369 @@
...
+            var compilation = CreatePatternCompilation(source);
+            compilation.VerifyDiagnostics(
+                );
+            var tree = compilation.SyntaxTrees[0];
+            var model = compilation.GetSemanticModel(tree);
+
+            var subpatterns = tree.GetRoot().DescendantNodes().OfType<SubpatternSyntax>().ToArray();
+            Assert.Equal(2, subpatterns.Length);
+
+            AssertEmpty(model.GetSymbolInfo(subpatterns[0]));
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests3.cs]: >AssertEmpty(model.GetSymbolInfo(subpatterns[0])); [](start = 12, length = 49)

Is this behavior (`GetSymbolInfo` only returning a result on the name identifier, but not on the name-colon) consistent with the behavior on invocation arguments or tuple element names?

File: src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests3.cs [hunk 3]
```diff
@@ -0,0 +1,369 @@
...
+        {
+            var source =
+@"
+using System;
+class Program
+{
+    public static void Main()
+    {
+        Point p = null;
+        Console.WriteLine(p is ( X: 3, Y: 4 ));
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests3.cs]: >( X: 3, Y: 4 ) [](start = 31, length = 14)

Nit: the formatting for positional sub-patterns is atypical (here and elsewhere). `p is (X: 3, Y: 4)`

File: src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests3.cs [hunk 4]
```diff
@@ -0,0 +1,369 @@
...
+
+            var subpatterns = tree.GetRoot().DescendantNodes().OfType<SubpatternSyntax>().ToArray();
+            Assert.Equal(2, subpatterns.Length);
+
+            AssertEmpty(model.GetSymbolInfo(subpatterns[0]));
+            AssertEmpty(model.GetSymbolInfo(subpatterns[0].NameColon));
+            var x = subpatterns[0].NameColon.Name;
+            var xSymbol = model.GetSymbolInfo(x);
+            Assert.Equal(CandidateReason.None, xSymbol.CandidateReason);
+            Assert.Null(xSymbol.Symbol);
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests3.cs]: >Null [](start = 19, length = 4)

Why is the behavior with a tuple using different names different than the behavior with a Deconstruct using different names (covered by `PositionalPatternSymbolInfo_02`)?

File: src/Compilers/CSharp/Portable/Syntax/SyntaxFactory.cs [hunk 5]
```diff
@@ -2161,6 +2161,14 @@ internal static CSharpSyntaxNode GetStandaloneNode(CSharpSyntaxNode node)
                         return parent;
                     }
 
+                    break;
+
+                case SyntaxKind.NameColon:
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Syntax/SyntaxFactory.cs]: >NameColon [](start = 32, length = 9)

I didn't understand this. A subpattern is not an expression, but this method should return "the containing expression that is actually a language expression".

File: src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs [hunk 6]
```diff
@@ -1853,6 +1853,7 @@ internal protected virtual CSharpSyntaxNode GetBindableSyntaxNode(CSharpSyntaxNo
                 case SyntaxKind.OperatorDeclaration:
                 case SyntaxKind.ConversionOperatorDeclaration:
                 case SyntaxKind.GlobalStatement:
+                case SyntaxKind.Subpattern:
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs]: >Subpattern [](start = 32, length = 10)

I didn't understand this either. I'd have expected we needed the containing expression to bind the sub-pattern.
For instance, in `e is { X = 3 }` we need to bind the whole thing including `e`. Or in recursive patterns.

File: src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs [hunk 7]
```diff
@@ -1844,6 +1849,16 @@ private bool IsInTypeofExpression(int position)
             return SymbolInfoFactory.Create(symbols, resultKind, isDynamic);
         }
 
+        private SymbolInfo GetSymbolInfoForSubpattern(BoundSubpattern subpattern)
+        {
+            if (subpattern.Symbol?.OriginalDefinition is ErrorTypeSymbol originalErrorType)
+            {
+                return new SymbolInfo(null, originalErrorType.CandidateSymbols.CastArray<ISymbol>(), originalErrorType.ResultKind.ToCandidateReason());
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs]: Nit: consider naming the `null` argument

**Comment by jcouv** [comment]
>         /// Get the semantic info of a named argument in an invocation-like expression.

The comment seems out-of-date

---
Refers to: src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs:4089 in 607639b. [](commit_id = 607639b7e061624d1cfe343fe25cdfa2c844bbe4, deletion_comment = False)

File: src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs [hunk 8]
```diff
@@ -1756,6 +1756,14 @@ private bool IsInTypeofExpression(int position)
             BoundNode boundNodeForSyntacticParent,
             Binder binderOpt)
         {
+            LookupResultKind resultKind;
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs]: Nit: Those two locals can be declared further down.

File: src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs [hunk 9]
```diff
@@ -554,29 +554,27 @@ private BoundPattern BindRecursivePattern(RecursivePatternSyntax node, TypeSymbo
...
-                        BindPattern(subPattern.Pattern, elementType, isError, diagnostics));
+                        BindPattern(subpatternSyntax.Pattern, elementType, isError, diagnostics));
                     patterns.Add(boundSubpattern);
                 }
             }
             else
             {
                 // It is not a tuple type. Seek an appropriate Deconstruct method.
                 var inputPlaceholder = new BoundImplicitReceiver(node, declType); // A fake receiver expression to permit us to reuse binding logic
-                // PROTOTYPE(patterns2): Can we include element names node.Subpatterns[i].NameColon?.Name in the AnalyzedArguments
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs]: >PROTOTYPE [](start = 19, length = 9)

I assume we have a test for this scenario (two Deconstruct methods, deconstruction pattern is ambiguous despite names).
Did LDM make a decision on this (in the scope of C# 8.0)?

**Comment by jcouv** [comment]
Looks like an IDE test needs updating: `GenerateVariableTests.TestPropertyPatternGenerateConstant`.

File: src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests3.cs (hunk 2 above)
Comment (gafter) [reply src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests3.cs]: Yes. `GetSymbolInfo` is defined on expressions and a couple of other syntactic constructs. The name in a name-colon is an expression, but the enclosing name-colon isn't.

---
In reply to: [191019103](https://github.com/dotnet/roslyn/pull/27160#discussion_r191019103) [](ancestors = 191019103)

File: src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests3.cs (hunk 4 above)
Comment (gafter) [reply src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests3.cs]: In a `Deconstruct`, we can tell what symbol is uniquely identified by the identifier merely by the position. If the name is incorrect, it is and error, but it still refers to that (positional) parameter.

In a tuple, there may be more than one name for a given tuple position (e.g. `x` and `Item1`). So in that case we use name lookup to decide what element field was intended.

---
In reply to: [191020864](https://github.com/dotnet/roslyn/pull/27160#discussion_r191020864) [](ancestors = 191020864)

File: src/Compilers/CSharp/Portable/Syntax/SyntaxFactory.cs (hunk 5 above)
Comment (gafter) [reply src/Compilers/CSharp/Portable/Syntax/SyntaxFactory.cs]: Will adjust comment. It is not just for expressions, it is for anything that can be used for GetSymbolInfo. That now includes a subpattern.

---
In reply to: [191021322](https://github.com/dotnet/roslyn/pull/27160#discussion_r191021322) [](ancestors = 191021322)

File: src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs (hunk 6 above)
Comment (gafter) [reply src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs]: The name of this method is perhaps poorly chosen. It returns the syntax that we'd find on a bound node of interest in the map. To get the node in the map we would indeed have to bind the whole statement.

---
In reply to: [191021984](https://github.com/dotnet/roslyn/pull/27160#discussion_r191021984) [](ancestors = 191021984)

File: src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs (hunk 9 above)
Comment (gafter) [reply src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs]: This is an open issue in [https://github.com/dotnet/csharplang/issues/1054](https://github.com/dotnet/csharplang/issues/1054)

---
In reply to: [191023469](https://github.com/dotnet/roslyn/pull/27160#discussion_r191023469) [](ancestors = 191023469)

**Comment by gafter** [comment]
@jcouv I think all of your comments have been addressed.
@agocke Can you please have a look at this PR?

**Comment by gafter** [comment]
@jcouv Do you have any other comments?

**Pull request merged**